File: TwoPointer/3.longest-substring-without-repeating-characters.go
package leetcode

// @leet start
func lengthOfLongestSubstring(s string) int {
	set := make(map[rune]bool)
	l, ans := 0, 0

	for r, char := range s {
		for set[char] {
			delete(set, rune(s[l]))
			l++
		}
		set[char] = true
		if r-l+1 > ans {
			ans = r - l + 1
		}
	}

	return ans
}

// @leet end

File: TwoPointer/3.longest-substring-without-repeating-characters.ts
// @leet start
function lengthOfLongestSubstring(s: string): number {
  const set = new Set();
  let l = 0;
  let ans = 0;
  for (let r = 0; r < s.length; r++) {
    while (set.has(s[r])) {
      set.delete(s[l]);
      l++;
    }
    set.add(s[r]);
    ans = Math.max(ans, r - l + 1);
  }
  return ans;
}
// @leet end

